dialogue-okhttp-client/src/main/java/com/palantir/dialogue/OkHttpChannels.java

```diff
@@ -0,0 +1,161 @@
+    private static final boolean DEFAULT_ENABLE_HTTP2 = true;
+
+    private static final ThreadFactory executionThreads = new ThreadFactoryBuilder()
+            .setUncaughtExceptionHandler((thread, uncaughtException) -> log.error(
+                    "An exception was uncaught in an execution thread. "
+                            + "This likely left a thread blocked, and is as such a serious bug "
+                            + "which requires debugging.",
+                    uncaughtException))
+            .setNameFormat("remoting-okhttp-dispatcher-%d")
+            // This diverges from the OkHttp default value, allowing the JVM to cleanly exit
+            // while idle dispatcher threads are still alive.
+            .setDaemon(true)
+            .build();
+
+    /**
+     * The {@link ExecutorService} used for the {@link Dispatcher}s of all OkHttp clients created through this class.
+     * Similar to OkHttp's default, but with two modifications:
+     *
+     * <ol>
+     *   <li>A logging uncaught exception handler
+     *   <li>Daemon threads: active request will not block JVM shutdown <b>unless</b> another non-daemon thread blocks
+     *       waiting for the result. Most of our usage falls into this category. This allows JVM shutdown to occur
+     *       cleanly without waiting a full minute after the last request completes.
+     * </ol>
+     */
+    private static final ExecutorService executionExecutor = Executors.newCachedThreadPool(executionThreads);
+
+    /** Shared dispatcher with static executor service. */
+    private static final Dispatcher dispatcher;
+
+    static {
+        dispatcher = new Dispatcher(executionExecutor);
+        // Restricting concurrency is done elsewhere in ConcurrencyLimiters.
+        dispatcher.setMaxRequests(Integer.MAX_VALUE);
+        // Must be less than maxRequests so a single slow host does not block all requests
+        dispatcher.setMaxRequestsPerHost(256);
+    }
+
+    /** Shared connection pool. */
+    private static final ConnectionPool connectionPool = new ConnectionPool(
+            1000,
+            // Most servers use a one minute keepalive for idle connections, by using a shorter keepalive on
+            // clients we can avoid race conditions where the attempts to reuse a connection as the server
+            // closes it, resulting in unnecessary I/O exceptions and retrial.
+            55,
+            TimeUnit.SECONDS);
+
+    private OkHttpChannels() {}
+
+    public static Channel create(ClientConfiguration config, UserAgent baseAgent, TaggedMetricRegistry metrics) {
+        OkHttpClient.Builder builder = new OkHttpClient()
+                .newBuilder()
+                .dispatcher(dispatcher)
+                .connectionPool(connectionPool)
+                .followRedirects(false) // We implement our own redirect logic.
+                .sslSocketFactory(config.sslSocketFactory(), config.trustManager())
```

Comment:
have we lost the KeepAliveSslSocketFactory here? Maybe we should just make it public in c-j-r?

Reply:
That's correct, this doesn't record client tls handshake metrics or set keepalive. Agree that we should be consistent across all clients.